test/e2e/networking_perf.go

```diff
@@ -0,0 +1,249 @@
+var _ = KubeDescribe("Networking IPerf [Experimental] [Feature:Network-Performance] ", func() {
+
+	f := NewDefaultFramework("network-perf")
+
+	// A few simple bandwidth tests which are capped by nodes.
+	runClientServerBandwidthMeasurement(f, 1, gceBandwidthBitsEstimate)
+	runClientServerBandwidthMeasurement(f, 5, gceBandwidthBitsEstimate)
+	runClientServerBandwidthMeasurement(f, 8, gceBandwidthBitsEstimate)
+
+	// As many clients as possible based on node count.
+	runClientServerBandwidthMeasurement(f, 10000, gceBandwidthBitsEstimate)
+})
+
+func runClientServerBandwidthMeasurement(f *Framework, numClient int, maxBandwidthBits int64) {
+
+	// TODO: Make this a function parameter, once we distribute iperf endpoints.
+	numServer := 1
+
+	It(fmt.Sprintf("should transfer ~ 1GB onto the service endpoint %v servers (maximum of %v clients)", numServer, numClient), func() {
+		nodes := ListSchedulableNodesOrDie(f.Client)
+		totalPods := len(nodes.Items)
+		// for a single service, we expect to divide bandwidth between the network.  Very crude estimate.
+		expectedBandwidth := int(float64(maxBandwidthBits) / float64(totalPods))
+		Expect(totalPods).NotTo(Equal(0))
+		appName := "iperf-e2e"
+		CreateServiceWithPods(f,
+			8001,
+			8002,
+			appName,
+			api.Container{
+				Name:  "iperf-server",
+				Image: "mattgruter/iperf",
+				Args: []string{
+					"-s", "-p", "8001"},
+				Ports: []api.ContainerPort{{ContainerPort: 8001}},
+			},
+			// this will be used to generate the -service name which all iperf clients point at.
+			numServer, // 1 server
+		)
+		CreateServiceWithPods(f,
+			-1,
+			-1,
+			"iperf-e2e-cli", // this will be used to generate the -service name which all iperf clients point at.
+			api.Container{
+				Name:  "iperf-client",
+				Image: "mattgruter/iperf",
+				Args: []string{
+					"-c", "service-for-" + appName, "-p", "8002", "--reportstyle", "C",
+				},
+			},
+			numClient, // 1 server
+		)
+		Logf("Reading all perf results to stdout.")
+		Logf("date,cli,cliPort,server,serverPort,id,interval,transferBits,bandwidthBits")
+		bandwidthMap := make(map[string]int64)
+
+		// Calculate expected number of clients based on total nodes.  ITs own function since this
+		// logic may change over time at scale.
+		expectedCli := func() int {
+			nodes, err := GetReadyNodes(f)
+			expectNoError(err)
+			return int(math.Min(float64(len(nodes.Items)), float64(numClient)))
+		}()
+
+		start := time.Now()
+		for len(bandwidthMap) < expectedCli && time.Since(start) <= 2*time.Minute {
+			TryEachPod(f.Client, f.Namespace.Name, "app", "iperf-e2e-cli-pod",
+				func(p api.Pod) {
+					resultS, err := lookForStringInLog(f.Namespace.Name, p.Name, "iperf-client", "0-10.0", 1*time.Second)
+					if err == nil {
+						Logf(resultS)
+						ipr := NewIPerf(resultS)
+						bandwidthMap[p.Name+"-"+ipr.cli+"-"+ipr.id+"--server="+ipr.server] = ipr.bandwidthBits
```

Comment:
Why not create a IPerfResults struct which is this map + a util to translate the IPerfResult, as that would make is easier to read.   Plus the logic below could go into an add f(n)